File: system-design/oop-patterns/04-factory-method-pattern/full/java/zoo/BalancedAnimalFactory.java
package zoo;

public class BalancedAnimalFactory implements AnimalFactory {
    private int dogCount = 0;
    private int catCount = 0;
    private int duckCount = 0;

    public Animal createAnimal() {
        if (dogCount <= catCount && dogCount <= duckCount) {
            dogCount++;
            return new Dog();
        }
        if (catCount <= duckCount) {
            catCount++;
            return new Cat();
        }
        duckCount++;
        return new Duck();
    }
}

File: system-design/oop-patterns/04-factory-method-pattern/full/js/zoo/BalancedAnimalFactory.js
import { Dog } from './Dog.js';
import { Cat } from './Cat.js';
import { Duck } from './Duck.js';

export class BalancedAnimalFactory {
  #dogCount = 0;
  #catCount = 0;
  #duckCount = 0;

  createAnimal() {
    if (this.#dogCount <= this.#catCount && this.#dogCount <= this.#duckCount) {
      this.#dogCount++;
      return new Dog();
    }
    if (this.#catCount <= this.#duckCount) {
      this.#catCount++;
      return new Cat();
    }
    this.#duckCount++;
    return new Duck();
  }
}
